src/System.Net.Security/src/System/Net/SecureProtocols/SslStream.cs

```diff
@@ -0,0 +1,454 @@
+        }
+
+        public virtual Task AuthenticateAsClientAsync(string targetHost, X509CertificateCollection clientCertificates, SslProtocols enabledSslProtocols, bool checkCertificateRevocation)
+        {
+            return Task.Factory.FromAsync((callback, state) => BeginAuthenticateAsClient(targetHost, clientCertificates, enabledSslProtocols, checkCertificateRevocation, callback, state), EndAuthenticateAsClient, null);
+        }
+
+        public virtual Task AuthenticateAsServerAsync(X509Certificate serverCertificate)
+        {
+            return Task.Factory.FromAsync(BeginAuthenticateAsServer, EndAuthenticateAsServer, serverCertificate, null);
+        }
+
+        public virtual Task AuthenticateAsServerAsync(X509Certificate serverCertificate, bool clientCertificateRequired, SslProtocols enabledSslProtocols, bool checkCertificateRevocation)
+        {
+            return Task.Factory.FromAsync((callback, state) => BeginAuthenticateAsServer(serverCertificate, clientCertificateRequired, enabledSslProtocols, checkCertificateRevocation, callback, state), EndAuthenticateAsServer, null);
+        }
+        #endregion
+
+        public override bool IsAuthenticated
+        {
+            get
+            {
+                return _SslState.IsAuthenticated;
+            }
+        }
+
+        public override bool IsMutuallyAuthenticated
+        {
+            get
+            {
+                return _SslState.IsMutuallyAuthenticated;
+            }
+        }
+
+        public override bool IsEncrypted
+        {
+            get
+            {
+                return IsAuthenticated;
+            }
+        }
+
+        public override bool IsSigned
+        {
+            get
+            {
+                return IsAuthenticated;
+            }
+        }
+
+        public override bool IsServer
+        {
+            get
+            {
+                return _SslState.IsServer;
+            }
+        }
+
+        public virtual SslProtocols SslProtocol
+        {
+            get
+            {
+                return _SslState.SslProtocol;
+            }
+        }
+
+        public virtual bool CheckCertRevocationStatus
+        {
+            get
+            {
+                return _SslState.CheckCertRevocationStatus;
+            }
+        }
+
+        public virtual X509Certificate LocalCertificate
+        {
+            get
+            {
+                return _SslState.LocalCertificate;
+            }
+        }
+
+        public virtual X509Certificate RemoteCertificate
+        {
+            get
+            {
+                _SslState.CheckThrow(true);
+
+                object chkCertificateOrBytes = _remoteCertificateOrBytes;
+                if (chkCertificateOrBytes != null && chkCertificateOrBytes.GetType() == typeof(byte[]))
+                    return (X509Certificate)(_remoteCertificateOrBytes = new X509Certificate((byte[])chkCertificateOrBytes));
```

Comment:
I assume the return type is locked to X509Certificate(1) because of legacy constraints, but why only allocate it as an X509Certificate(1) object instead of X509Certificate2?